In [1]:
import pandas as pd

In [2]:
mt=pd.read_csv('MT_cleaned.csv',delimiter=',')
vt=pd.read_csv('VT_cleaned.csv',delimiter=',')

F:\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (25,27) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
F:\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (19) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
mt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825118 entries, 0 to 825117
Data columns (total 33 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     825118 non-null  object 
 1   state                  825118 non-null  object 
 2   stop_date              825107 non-null  object 
 3   stop_time              825107 non-null  object 
 4   location_raw           825114 non-null  object 
 5   county_name            821062 non-null  object 
 6   county_fips            821062 non-null  float64
 7   fine_grained_location  821377 non-null  object 
 8   police_department      0 non-null       float64
 9   driver_gender          824999 non-null  object 
 10  driver_age_raw         825118 non-null  float64
 11  driver_age             821638 non-null  float64
 12  driver_race_raw        825012 non-null  object 
 13  driver_race            822379 non-null  object 
 14  violation_raw          825045 non-nu

In [4]:
mt.columns

Index(['id', 'state', 'stop_date', 'stop_time', 'location_raw', 'county_name',
       'county_fips', 'fine_grained_location', 'police_department',
       'driver_gender', 'driver_age_raw', 'driver_age', 'driver_race_raw',
       'driver_race', 'violation_raw', 'violation', 'search_conducted',
       'search_type_raw', 'search_type', 'contraband_found', 'stop_outcome',
       'is_arrested', 'lat', 'lon', 'ethnicity', 'city', 'out_of_state',
       'vehicle_year', 'vehicle_make', 'vehicle_model', 'vehicle_style',
       'search_reason', 'stop_outcome_raw'],
      dtype='object')

1.The proportion of traffic stops in MT involving male drivers

In [5]:
male_mt=mt[mt['driver_gender']=='M']

In [6]:
male_mt.head()

id state   stop_date stop_time location_raw         county_name  \
1  MT-2009-00002    MT  2009-01-02     11:34     MISSOULA     Missoula County   
2  MT-2009-00003    MT  2009-01-03     11:36     MISSOULA     Missoula County   
4  MT-2009-00005    MT  2009-01-04     10:46     MISSOULA     Missoula County   
5  MT-2009-00006    MT  2009-01-04     14:41      RAVALLI      Ravalli County   
6  MT-2009-00007    MT  2009-01-04     17:45  YELLOWSTONE  Yellowstone County   

   county_fips        fine_grained_location  police_department driver_gender  \
1      30063.0  HWY 93 SO AND ANNS LANE S/B                NaN             M   
2      30063.0        P007 HWY 93 MM 77 N/B                NaN             M   
4      30063.0        P007 HWY 93 MM 81 N/B                NaN             M   
5      30081.0        P007 HWY 93 MM 67 S/B                NaN             M   
6      30111.0        WESTBOUND TRUCK SCALE                NaN             M   

   ...         lon  ethnicity city out_of_state vehicle_year vehicle_make  \
1  ... -114.081142          N  NaN        False         1996          GMC   
2  ... -114.073505          N  NaN        False         1999          GMC   
4  ... -114.079150        NaN  NaN        False         1992         TOYT   
5  ... -114.110712        NaN  NaN        False         1998         JEEP   
6  ... -108.342832          N  NaN        False         2006         SUBA   

   vehicle_model  vehicle_style search_reason            stop_outcome_raw  
1             TK          TRUCK           NaN  INFFRACTION ARREST,WARNING  
2          YUKON  SPORT UTILITY           NaN          INFFRACTION ARREST  
4         TERCEL          SEDAN           NaN          INFFRACTION ARREST  
5            CHE  SPORT UTILITY           NaN  INFFRACTION ARREST,WARNING  
6            NaN             PK           NaN            TRAFFIC CITATION  

[5 rows x 33 columns]

In [7]:
male_mt.tail()

id state stop_date stop_time location_raw         county_name  \
825113  MT-NA-07    MT       NaN       NaN       DAWSON       Dawson County   
825114  MT-NA-08    MT       NaN       NaN    ROOSEVELT    Roosevelt County   
825115  MT-NA-09    MT       NaN       NaN  YELLOWSTONE  Yellowstone County   
825116  MT-NA-10    MT       NaN       NaN       DAWSON       Dawson County   
825117  MT-NA-11    MT       NaN       NaN       CARBON       Carbon County   

        county_fips                         fine_grained_location  \
825113      30021.0           CAINS COULEE ROAD X RADIO HILL ROAD   
825114      30085.0             BIA ROUTE 1 WEST FORT KIPP 1 MILE   
825115      30111.0                    STATE AVE. AND S. 27TH ST.   
825116      30021.0                                MT 16 MM 10 SB   
825117      30009.0  MP 17 HIGHWAY 72 NB / .1 NORTH OF PRICE ROAD   

        police_department driver_gender  ...         lon  ethnicity  \
825113                NaN             M  ... -104.734623          N   
825114                NaN             M  ... -104.785495          N   
825115                NaN             M  ... -108.492856          N   
825116                NaN             M  ... -104.646874          N   
825117                NaN             M  ... -108.964528          N   

                      city out_of_state vehicle_year     vehicle_make  \
825113  NOT IN CITY LIMITS        False         1998            PONTI   
825114  NOT IN CITY LIMITS         True         2011            DODGE   
825115            BILLINGS        False         1995  CHEVROLET (CHEV   
825116            GLENDIVE        False         2004     DODGE (DODG)   
825117                 NaN         True         2007           TOYOTa   

        vehicle_model  vehicle_style search_reason            stop_outcome_raw  
825113     GRAND PRIX          SEDAN           NaN                     WARNING  
825114       RAM 1500         PICKUP           NaN                     WARNING  
825115            C15         Pickup           NaN    TRAFFIC CITATION,WARNING  
825116        DURANGO  SPORT UTILITY           NaN  MISDEMEANOR ARREST,WARNING  
825117         matrix          SEDAN           NaN                     WARNING  

[5 rows x 33 columns]

In [8]:
male_driver=len(male_mt)

In [9]:
prop=male_driver/len(mt)
prop

0.6749749732765495

the proportion: 0.6749749732765495

2.Factor increase in a traffic stop arrest likelihood in MT from OOS plates

3.The proportion of traffic stops in MT involving speeding violations

In [10]:
mt['violation'].dropna

<bound method Series.dropna of 0         Other,Paperwork,Safe movement
1          Other (non-mapped),Seat belt
2                              Speeding
3                              Speeding
4                    Seat belt,Speeding
                      ...              
825113      License,Other,Safe movement
825114                         Speeding
825115             DUI,Lights,Paperwork
825116           License,Other,Speeding
825117          License,Other,Paperwork
Name: violation, Length: 825118, dtype: object>

In [11]:
v=mt['violation']
mt_speed_violation=v.str.contains('Speeding',regex='True')
mt_speed_violation

0         False
1         False
2          True
3          True
4          True
          ...  
825113    False
825114     True
825115    False
825116     True
825117    False
Name: violation, Length: 825118, dtype: object

In [12]:
speed_violation_data_frame=pd.DataFrame(mt_speed_violation)

In [13]:
mt_speed_violation_filtered=speed_violation_data_frame.loc[speed_violation_data_frame['violation'] == True]

In [14]:
len(mt_speed_violation_filtered)/len(mt)

0.6580998111785223

The proportion of traffic stops in MT involving speeding violations: 0.6580998111785223

4.Factor increase in traffic stop DUI likelihood in MT over VT:

5.The average manufacture year of vehicles stopped in MT in 2020

In [15]:
#stop_mt=mt['stop_date'].dropna()
manufacture_year=mt[['vehicle_year','stop_date']]
manufacture_year

vehicle_year   stop_date
0              1994  2009-01-01
1              1996  2009-01-02
2              1999  2009-01-03
3              2002  2009-01-04
4              1992  2009-01-04
...             ...         ...
825113         1998         NaN
825114         2011         NaN
825115         1995         NaN
825116         2004         NaN
825117         2007         NaN

[825118 rows x 2 columns]

In [16]:
manufacture_year['stop_date']= pd.to_datetime(manufacture_year['stop_date'])

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [17]:
#extract year
manufacture_year['stop_year']=manufacture_year['stop_date'].apply(lambda x: x.year)

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [18]:
manufacture_year.dropna()

vehicle_year  stop_date  stop_year
0              1994 2009-01-01     2009.0
1              1996 2009-01-02     2009.0
2              1999 2009-01-03     2009.0
3              2002 2009-01-04     2009.0
4              1992 2009-01-04     2009.0
...             ...        ...        ...
825102         1995 2016-12-31     2016.0
825103         1999 2016-12-31     2016.0
825104         1993 2016-12-31     2016.0
825105         1996 2016-12-31     2016.0
825106         1997 2016-12-31     2016.0

[818497 rows x 3 columns]

In [19]:
manufacture_year.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825118 entries, 0 to 825117
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   vehicle_year  818508 non-null  object        
 1   stop_date     825107 non-null  datetime64[ns]
 2   stop_year     825107 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 18.9+ MB


In [ ]:
vehicle_year=pd.DataFrame(manufacture_year['vehicle_year'])
vehicle_year_avg=vehicle_year.astype(str).mean()
vehicle_year_avg

6.The difference in the total number of stops that occurred between min and max hours in both MT and VT

In [20]:
mt.stop_time.dropna()

0         02:10
1         11:34
2         11:36
3         10:33
4         10:46
          ...  
825102    23:54
825103    23:55
825104    23:58
825105    23:58
825106    23:58
Name: stop_time, Length: 825107, dtype: object

In [21]:
stop_time=pd.DataFrame(mt.stop_time.dropna())

In [22]:
stop_time.head()

stop_time
0     02:10
1     11:34
2     11:36
3     10:33
4     10:46

7.The area in sq. km of the largest county in MT 


In [ ]:
mt.iloc[105]

In [ ]:
mt['county_name'].describe()